keps/sig-storage/2639-secret-protection/README.md

```diff
@@ -0,0 +1,450 @@
+- Protect resources other than secrets from being deleted.
+
+## Proposal
+
+A new controller to protect secret is introduced.
+
+### User Stories
+
+#### Story 1
+
+A user creates a secret and a pod using the secret. Then, the user mistakenly delete the secret while the pod is using it.
+The secret is protected until the pod using the secret is deleted.
+
+#### Story 2
+
+A user creates a volume that uses a certain secret in the same namespace. Then, the user delete the namespace.
+The secret is protected until the volume using the secret is deleted and the deletion of the volume succeeds.
+
+#### Story 3
+
+A user really would like to delete a secret despite that it is used by other resources.
+The user force delete the secret while it is used by other resources, and the secret isn't protected and is actually deleted.
+
+### Notes/Constraints/Caveats (Optional)
+
+- Compatibility:
+  - There might be many existing scripts that don't care the order of deletion. Therefore, such scripts might stuck on secret deletion, if the deletion of the resources using the secrets are done later.
+- Usability:
+  - Use of the secret in other resource will not be obvious to users. Therefore, users might not easily understand why the secret is not deleted.
+  - Users might need to force delete the secret on deletion or would like to avoid protection for certain secrets that already exist or that are newly created.
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+This feature would be able to implement in the same way as `pv-protection/pvc-protection`.
+Protection logics for in-tree resources and out-of-tree resources are separeted and independently work.
+It is due to the restriction that CRDs can't be handled from in-tree controller.
+
+- In-tree resources(`Pod` and `PersistentVolume`):
+  - The deletion is blocked by using newly introduced `kubernetes.io/secret-protection` finalizer,
+  - The `kubernetes.io/secret-protection` finalizer will be added on creation of the secret by using admission controller for in-tree resources,
+  - The `kubernetes.io/secret-protection` finalizer will be deleted by newly introduced in-tree `secret-protection-controller` by checking whether the secret is in-use, on every change(Create/Update/Delete) events for secrets and related resources.
+- Out-of-tree resources(`VolumeSnapshot`):
+  - The deletion is blocked by using newly introduced `snapshot.storage.kubernetes.io/secret-protection` finalizer,
+  - The `snapshot.storage.kubernetes.io/secret-protection` finalizer will be added on creation of the secret by using admission controller for `VolumeSnapshot`,
```

Comment:
this seems better to do with a controller that adds the finalizer as a precondition to using the secret, rather than putting a blocking admission plugin in the secret-writing path

Reply:
@mkimuram have you addressed this comment?

Reply:
@xing-yang 

Sorry. I missed to respond to it. This is the same approach as the existing [pv/pvc protection](https://github.com/kubernetes/kubernetes/blob/master/plugin/pkg/admission/storage/storageobjectinuseprotection/admission.go). This logic is introduced [here](https://github.com/kubernetes/kubernetes/pull/55824/commits/a06901a8687e2270215cb98ca45f1d50b237c748). So, the discussion should be [here](https://github.com/kubernetes/kubernetes/issues/45143#issuecomment-335846191) or [here](https://github.com/kubernetes/community/pull/1174), however I can't find the reason why admission plugin is used there.

Actually, it has a fallback logic to add finalizer [here](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/volume/pvcprotection/pvc_protection_controller.go#L184) in pvc protection controller side. I'm not sure that this works well in all the cases. Therefore, I think that it's safer to use the same approach that worked for a long time.

@liggitt 

Do you know the background or do you think that the only fallback logic is enough? (It seems that you are involved in the discussion.)
I would like to make this KEP as is and remove this later if it really proves to be unnecessary. Does it make sense?